pkg/erofs/erofs.go

```diff
@@ -0,0 +1,691 @@
+	if i.DataLayout() != INODE_DATALAYOUT_FLAT_PLAIN {
+		return 0, linuxerr.ENOTSUP
+	}
+	return i.dataOffset, nil
+}
+
+// Data returns the read-only file data of this inode. Only "flat file data
+// without data inline (no extent)" data layout is supported.
+func (i *Inode) Data() ([]byte, error) {
+	if i.DataLayout() != INODE_DATALAYOUT_FLAT_PLAIN {
+		return nil, linuxerr.ENOTSUP
+	}
+	return i.image.bytes[i.dataOffset : i.dataOffset+i.size], nil
+}
+
+// hasInlineData indicates whether this inode has any inline data.
+func (i *Inode) hasInlineData() bool {
+	if i.DataLayout() != INODE_DATALAYOUT_FLAT_INLINE {
+		return false
+	}
+	blockSize := i.image.BlockSize()
+	if blockSize-uint32(i.inodeSize) >= uint32(i.size)&(blockSize-1) {
+		return true
+	}
+	return false
+}
+
+// populateChildMapCache populates the cache of the mappings of child names to inodes.
+//
+// Preconditions:
+//   - i.IsDir().
+func (i *Inode) populateChildMapCache() (map[string]*Inode, error) {
+	i.dirMu.Lock()
+	defer i.dirMu.Unlock()
+
+	if i.childMap != nil {
+		return i.childMap, nil
+	}
+
+	childMap := make(map[string]*Inode)
+
+	start := i.dataOffset
+	if i.hasInlineData() && i.blocks == 1 {
+		start = i.idataOffset
+	}
+
+	blockSize := uint64(i.image.BlockSize())
+	blocks := i.blocks
+	for blocks > 0 {
+		currOff := start
+		d := &Dirent{}
+		d.UnmarshalUnsafe(i.image.bytes[currOff:])
+		nameOff0 := start + uint64(d.NameOff)
+		maxSize := uint64(i.image.BlockSize())
+		if blocks == 1 {
+			maxSize = i.size % blockSize
+		}
+		for d != nil {
+			var (
+				next *Dirent
+				size uint64
+			)
+			nextOff := currOff + DIRENT_SIZE
+			if nextOff < nameOff0 {
+				next = &Dirent{}
+				next.UnmarshalUnsafe(i.image.bytes[nextOff:])
+				size = uint64(next.NameOff - d.NameOff)
+			} else {
+				next = nil
+				size = maxSize - uint64(d.NameOff)
+			}
+
+			buf := make([]byte, size)
+			bufp := primitive.ByteSlice(buf)
+			bufp.UnmarshalBytes(i.image.bytes[start+uint64(d.NameOff):])
+			n := bytes.IndexByte(bufp, 0)
+			if n == -1 {
+				n = len(bufp)
+			}
+			name := string(bufp[:n])
+			if name != "." && name != ".." {
+				if inode, err := i.image.newInode(d.FileType, d.Nid); err != nil {
+					return nil, err
+				} else {
+					childMap[name] = inode
+				}
+			}
+
+			d = next
+			currOff = nextOff
+		}
+
+		blocks--
+
+		if i.hasInlineData() && blocks == 1 {
+			start = i.idataOffset
+		} else {
+			start += blockSize
+		}
+	}
+
+	i.childMap = childMap
+	return i.childMap, nil
+}
+
+// childMapCache returns the cached mappings of child names to inodes.
+func (i *Inode) childMapCache() map[string]*Inode {
+	i.dirMu.RLock()
+	defer i.dirMu.RUnlock()
+	return i.childMap
+}
+
+// ChildMap returns the mappings of child names to inodes.
+func (i *Inode) ChildMap() (map[string]*Inode, error) {
+	if !i.IsDir() {
+		return nil, linuxerr.ENOTDIR
+	}
+	if childMap := i.childMapCache(); childMap != nil {
+		return childMap, nil
+	}
+	return i.populateChildMapCache()
+}
```

Comment:
ChildMap returns `i.childMap`, which might be modified by the caller (without holding i.dirMu). I suggestion having something like this:

```
func (i *Inode) WithChildMap(cb func() error) error {
	if !i.IsDir() {
		return linuxerr.ENOTDIR
	}
	// Fast path.
	i.dirMu.RLock()
	if i.childMap != nil {
		defer i.dirMu.RUnlock()
		return cb(i.childMap)
	}
	i.dirMu.RUnlock()

	// Populate childMap.
	if err := i.ensureChildMap(); err := nil {
		return err
	}

	i.dirMu.RLock()
	defer i.dirMu.RUnlock()
	return cb(i.childMap)
}
```

Reply:
Thanks for the suggestion. It’s looks better. Will fix it in the next version.